go/acct/radius.go

```diff
@@ -0,0 +1,360 @@
+	if err != nil {
+		logError("RADIUSSecret: Switch not found" + err.Error())
+		return nil, nil, err
+	}
+
+	return []byte(switchInfo.Secret), context.WithValue(ctx, switchInfoKey, switchInfo), nil
+}
+
+type Error string
+
+func (e Error) Error() string { return string(e) }
+
+const PacketTooSmall = Error("radius: packet not at least 20 bytes long")
+const PacketInvalidLength = Error("radius: invalid packet length")
+
+func checkPacket(raw []byte) error {
+	if len(raw) < 20 {
+		return PacketTooSmall
+	}
+
+	length := int(binary.BigEndian.Uint16(raw[2:4]))
+	if length < 20 || length > radius.MaxPacketLength || len(raw) != length {
+		return PacketInvalidLength
+	}
+
+	return nil
+}
+
+func packetToMap(ctx context.Context, p *radius.Packet) map[string]interface{} {
+
+	attributes := make(map[string]interface{})
+	for i, attr := range p.Attributes {
+		if rfc2865.VendorSpecific_Type == i {
+			for _, vattrs := range attr {
+				id, vsa, err := radius.VendorSpecific(vattrs)
+				if err != nil {
+					log.LoggerWContext(ctx).Error(fmt.Sprintf("Unknown vendor id: %d", id))
+					continue
+				}
+
+				v := radiusDictionary.GetVendorByNumber(uint(id))
+				if v == nil {
+					log.LoggerWContext(ctx).Error(fmt.Sprintf("Unknown vendor id: %d", id))
+					continue
+				}
+
+				for len(vsa) >= 3 {
+					vsaTyp, vsaLen := vsa[0], vsa[1]
+					data := vsa[2:int(vsaLen)]
+					a := dictionary.AttributeByOID(v.Attributes, []int{int(vsaTyp)})
+					if a == nil {
+						continue
+					}
+
+					addAttributeToMap(attributes, a, radius.Attribute(data))
+					vsa = vsa[int(vsaLen):]
+				}
+			}
+		} else {
+			a := radiusDictionary.GetAttributeByOID([]int{int(i)})
+			if a == nil {
+				log.LoggerWContext(ctx).Error(fmt.Sprintf("Unknown Attribute: %d", int(i)))
+				continue
+			}
+
+			addAttributeToMap(attributes, a, attr[0])
+		}
+	}
+
+	return attributes
+}
+
+func addAttributeToMap(attributes map[string]interface{}, da *dictionary.Attribute, attr radius.Attribute) {
+	var item interface{} = nil
+	switch da.Type {
+	case dictionary.AttributeString:
+		item = radius.String(attr)
+	case dictionary.AttributeInteger:
+		i, err := radius.Integer(attr)
+		if err == nil {
+			item = i
+		}
+	case dictionary.AttributeInteger64:
+		i, err := radius.Integer64(attr)
+		if err == nil {
+			item = i
+		}
+	case dictionary.AttributeIPAddr:
+		i, err := radius.IPAddr(attr)
+		if err == nil {
+			item = i.String()
+		}
+	}
+
+	if item != nil {
+		if old, found := attributes[da.Name]; found {
+			switch old.(type) {
+			case []interface{}:
+				attributes[da.Name] = append(old.([]interface{}), item)
+			default:
+				attributes[da.Name] = []interface{}{old, item}
+			}
+		} else {
+			attributes[da.Name] = item
+		}
+	} else {
+		logWarn(fmt.Sprintf("Not handled %s\n", da.Name))
+	}
+}
+
+func logError(msg string) {
+	log.LoggerWContext(context.Background()).Error(msg)
+}
+
+func logWarn(msg string) {
+	log.LoggerWContext(context.Background()).Warn(msg)
```

Comment:
You should use a global context with a logger in it, otherwise you'll be creating a new logger everytime